Details for Directed Graph Case

Armadilloa16 opened the pull request:
As per #160 , and including to typo fixes and some rearrangement for clarity.
Thanks @kortschak for the suggested changes, PTAL?

kortschak commented on community/louvain.tex:
```diff
@@ -190,97 +190,278 @@ \section{Directed Graphs}
-\begin{equation}
-	Q^{(1)} = \frac{1}{2m}\sum_i\sum_j\left[ \left( B_{ij} + B_{ji} \right) - \gamma \frac{k_ik_j}{2m} \right] \delta(c_i,c_j).
-\end{equation}
+We will show that one of the two `generalisation of expressions' approachs is 
```
s/approachs/approaches/

Armadilloa16 replied:
Done. Such a minor change -- and this diff isn't very informative anyway, should I just squash it straight away?

kortschak replied:
There's no hurry.

kortschak commented:
Typo at https://github.com/gonum/graph/pull/161/files#diff-2f2eb9df1138efa51cc6f12e5c38ffb6R138 s/infact/in fact/

kortschak commented:
LGTM with typos fixed, but wait for @vladimir-ch.

@vladimir-ch the intention here is for me to implement the directed case in addition to the approach that we currently have which the work here demonstrates to be different.

kortschak commented:
Also the typo at L138.

Armadilloa16 commented:
L138: $A^*_{\alpha \alpha}$ to be non-zero, infact

whats the typo? is `infact' bad english?

kortschak commented:
Yup. 'infact' is not in fact a word.

Armadilloa16 commented:
Haha okie dokie fixed it. Better check my thesis for that one -- I'm going to have proofreading PTSD by the time I have submitted.

vladimir-ch commented:
I am a little confused in section 3 and having answers to my questions would surely help.
- The first paragraph of Section 3 says that we treat incoming and outgoing edges equally. What does it mean, that they have the same weight? Maybe not because what would make the graph directed and also it seems to me that it is assumed that in Section 3.1 and 3.2 that B_ij != B_ji. But then in Section 3.3 it is said that "...**still** treats incoming and outgoing edges as equally weighted..." and Eq. (30) confirms it.
- Since (27) is a different modularity function, it is not surprising that that approach is not equivalent to the other two. Is there any justification that makes (27) a better option in some sense? Maybe some motivation behind the modification could be included at the beginning of 3.3.

Some nitpicking:
- Give the document a title?
- Capitalize only the first character in (sub)section titles.
- L36,37: I would write the sentences in infinitive form (1. Find ... 2. Construct ...)
- L48: Add `\section{Undirected graphs}` and change the following two sections `sec:initialPass` and `sec:laterPasses` to subsections.
- L72: I would write it as `belongs to, i.e., $\alpha = c_a$.` and I would remove the following equation because it is redundant. Similarly with Eq. (38).
- L152: Something is missing at the beginning of the sentence. Also on L415.
- L210: s/simple/simply/
- L280: Eq. (29) is empty

A couple of side remarks and nitpicking that do not need to be acted upon:
- I find it confusing that `c_i` means the community to which the node i belongs rather than the i-th community. I would prefer a notation that distinguishes between the two.
- Q is a function but it is not immediately clear what are the variables (it's how the nodes are partitioned into communities). The definition of Q should be `Q(something) := ...`.
- It's a matter of taste but I like to use := for definitions. For example, on L79, the two = have not the same meaning, the first = defines a new symbol, the second is an equality.

I didn't check the formulas in section 4.

Armadilloa16 commented:
Hi @vladimir-ch , here is my attempt at answering your questions:

First, and I agree the wording is confusing, what do I mean when I 
say `treat incoming and outgoing edges equally'. Your right to assume 
I do _not_ mean the edge weights are equal -- B_ij != B_ji is certainly 
allowed. What I mean is that outgoing edges are treated as being 
equally important as incoming edges. Perhaps an example is the best
way of explaining -- in Eq. (20) I propose the construction of an undirected
graph as A_ij = B_ij + B_ji, which is `symmetric' in a sense, and this is 
what I mean by `treating incoming and outgoing edges equally', and due 
to this symmetry A_ij = A_ji, which is needed in order for the graph with 
A_ij edges to be undirected. One could construct a new graph with edge 
weights A_ij = 2*B_ij + B_ji, and this would be assymetric, not treating incoming
as the same as outgoing, and would also not be usable as an undirected 
graph in most cases.
Similarly, the alternative modularity function in Eq. (27) also `treats incoming 
and outgoing edges equally' in the sense that it is invarient to swapping the 
terms `incoming' for `outgoing' everywhere.

As for motivating this alternate modularity function -- I think this might be 
discussed in one of the relevant papers? Its been so long since I read them.
@kortschak do you remember the context for this?

Other miscellania:
 - What would a good title for the document be?
 - About the notation c_i -- similarly to the wording `treat incoming and outgoing
edges equally', I agree its confusing but can't think of an easy way of saying the 
same thing otherwise.
 - If we let c be a function, and c(i) be the community to which node i belongs, 
would this be less confusing? and then Q could be a function of c -- Q(c). 
 - := vs =, I don't really mind. Happy to go with whatever convention is (if there is 
one).

Anything I haven't mentioned here, I simply agree with and incorporated.

vladimir-ch commented:
Hi @Armadilloa16 (in the heat of reading your draft I'm forgetting good manners :-), thanks for a detailed answer. I see, if in 2.3 `...as equally weighted...` is replaced with `...as equally important...`, it makes much more sense.

> What would a good title for the document be?

Anything would work I think. What about simply "Louvain algorithm for undirected and directed graphs"?

> If we let c be a function, and c(i) be the community to which node i belongs, would this be less confusing? and then Q could be a function of c -- Q(c).

It would, at least for me. It seems that it would not complicate the notation elsewhere, except for `\delta(c_i,c_j)` which could be written as `\delta_{ij}` (or `\delta_{ij}(c)`) anyway.

> := vs =

It's really not important, we can leave it as is. If I have time, I can add `:` later.

LGTM. You can add the motivation for the modularity function later.

kortschak commented:
>Maybe some motivation behind the modification could be included at the beginning of 3.3.

For the record, I had some email discussions with @vtraag when I was sorting out the undirected case and making the directed->undirected shim. He wrote the book on this kind of thing (literally - there's a Springer book that's essentially his [thesis](http://dial.academielouvain.be/handle/boreal:134615)). During these discussions he pointed out that under some circumstances you might expect qualitative differences in the communities found by the two approaches, particularly when there is a large discrepancy between ingoing and outgoing links. I don't have a good intuition what this difference might be (I'm not sure he did either), but there are some cases in my work where it might have an impact and I want to check that.

Armadilloa16 commented:
Hey @vladimir-ch all good, maths is fun, and therefore engrossing! 
Thanks for all the constructive comments, I've incorporated all the 
suggestions, using `\delta_{ij}(c)` instead of `\delta(c_i,c_j)` and 
`c(i)` instead of `c_i`.

Also, @kortschak, if you run some comparisons between these 
two approaches I think varying `\gamma` might be of interest, as 
my intuition points me in the direction that the difference between 
the two approaches might be related to gamma in some intimate 
way. Notice, for example, that as `\gamma` gets small, so does the
expression in Eq. (29) (of the new version) -- the expression for 
`Q^{(3)} - Q^{(2)}'

vladimir-ch commented on community/louvain.tex:
```diff
@@ -7,15 +7,19 @@
 \end{equation}
 where $A_{ij}$ is the edge weight between nodes $i$ and $j$, 
-$c_i$ is the current community to which node $i$ belongs,
+$c(i)$ is the current community to which node $i$ belongs,
```
If I were to nitpick, I would say that since c(i) does not appear in the formula anymore, I would replace this line with a description of what c itself is (`$c$ is a partition of nodes into disjoint subsets or communities`) and I would put it after `where`.

vladimir-ch commented on community/louvain.tex:
```diff
@@ -25,16 +29,16 @@
+	1 & \text{if} \quad  c(i) = c(j) \\
 	0 & \text{otherwise} \end{array} \right ..
 \end{equation}
 
```
Explanation of the `c(i)` notation can be put here. The word `current` is not necessary anymore, since it is already implicitly contained in `c`.

vladimir-ch commented:
> ... maths is fun, and therefore engrossing!

Definitely :-)

I added my last nitpick, so I repeat my LGTM.

Armadilloa16 commented:
Cool, thanks @vladimir-ch. 

I incorporated your nitpicks, with one small nitpick of my 
own -- the word `disjoint` is redundant as it is already 
implied in the standard definition of the word `partition`. 

Also, as I was calling `\maketitle` with no author defined 
TeX was throwing a warning, so I added an `\author` 
definition. 

Should I squash the commit history now?

vladimir-ch commented:
> I incorporated your nitpicks, with one small nitpick of my 
own -- the word disjoint is redundant as it is already 
implied in the standard definition of the word partition.

Great, thanks.

> Should I squash the commit history now?

Yes, please.

Armadilloa16 commented:
Cool, thanks for all the feedback!

Oh, whoops. What have I done...  let me try to update to gonum/graph, I think I'm behind a couple of commits. I'm still getting the hang of git -- bear with me please! :)

kortschak commented:
Thank you for doing this.

Armadilloa16 commented:
Its been fun! Thanks for teaching me how to contribute! :)
